pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,430 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package storecache
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/thanos-io/thanos/pkg/cache"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+type CachingBucketConfig struct {
+	// Basic unit used to cache chunks.
+	ChunkBlockSize int64 `yaml:"chunk_block_size"`
+
+	// Maximum number of GetRange requests issued by this bucket for single GetRange call. Zero or negative value = unlimited.
+	MaxChunksGetRangeRequests int `yaml:"max_chunks_get_range_requests"`
+
+	// TTLs for various cache items.
+	ChunkObjectSizeTTL time.Duration `yaml:"chunk_object_size_ttl"`
+	ChunkBlockTTL      time.Duration `yaml:"chunk_block_ttl"`
+}
+
+func DefaultCachingBucketConfig() CachingBucketConfig {
+	return CachingBucketConfig{
+		ChunkBlockSize:            16000, // Equal to max chunk size.
+		ChunkObjectSizeTTL:        24 * time.Hour,
+		ChunkBlockTTL:             24 * time.Hour,
+		MaxChunksGetRangeRequests: 3,
+	}
+}
+
+// Bucket implementation that provides some caching features, using knowledge about how Thanos accesses data.
+type CachingBucket struct {
+	bucket objstore.Bucket
+	cache  cache.Cache
+
+	config CachingBucketConfig
+
+	logger log.Logger
+
+	requestedChunkBytes prometheus.Counter
+	cachedChunkBytes    prometheus.Counter
+	fetchedChunkBytes   prometheus.Counter
+	refetchedChunkBytes prometheus.Counter
+
+	objectSizeRequests prometheus.Counter
+	objectSizeHits     prometheus.Counter
+}
+
+func NewCachingBucket(b objstore.Bucket, c cache.Cache, chunks CachingBucketConfig, logger log.Logger, reg prometheus.Registerer) (*CachingBucket, error) {
+	if b == nil {
+		return nil, errors.New("bucket is nil")
+	}
+	if c == nil {
+		return nil, errors.New("cache is nil")
+	}
+
+	return &CachingBucket{
+		config: chunks,
+		bucket: b,
+		cache:  c,
+		logger: logger,
+
+		requestedChunkBytes: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_requested_chunk_bytes_total",
+			Help: "Total number of requested bytes for chunk data ",
```

Comment:
Could you add a dot (`.`) at the end of each `Help` missing it, to keep consistency with Thanos codebase, please?